src/components/lib/shm_bm/shm_bm.h

```diff
@@ -0,0 +1,271 @@
+	if (id == 0) return 0;
+	memset(*shm, 0, round_up_to_page(alloc));
+
+	// set nobj bits to free
+	__shm_bm_set_contig((word_t *) *shm, nobj);
+	return id;
+}
+
+static inline void * 
+__shm_bm_clt_alloc(shm_bm_t shm, shm_objid_t *objid, size_t objsz, unsigned int nobj)
+{
+	int     freebit, idx , offset, lz;
+	word_t  word; 
+	word_t *bm;
+
+	/* 
+	 * Find a free space; could be preempted. Find the next word in
+	 * the bitmap with a free bit and get its index in the bitmap, and 
+	 * the offset of the free bit in that word. Then try to atomically 
+	 * xchg the bitmap with the freebit set to 0
+	 */
+	bm = (word_t *) shm;
+	do {
+		idx = __shm_bm_next_free_word(bm, SHM_BM_BITS_TO_WORDS(nobj));
+		if (idx == -1) return 0;
+		word   = bm[idx];
+		lz     = __builtin_clzl(word); 
+		offset = SHM_BM_BITMAP_BLOCK - lz - 1;
+	} while (!cos_cas(bm + idx, word, word & ~(1ul << offset)));
+
+	freebit = lz + (idx * SHM_BM_BITMAP_BLOCK);
+
+	cos_faab(SHM_BM_REFC(shm, nobj) + freebit, 1);
+
+	*objid = (shm_objid_t) freebit;
+	return SHM_BM_DATA(shm, nobj) + (freebit * objsz);
+}
+
+static inline shm_reqtok_t
+__shm_bm_srv_map(struct shm_service_tbl_ent *srv_tbl, shm_reqtok_t *nxt_reqtok, cbuf_t shm_id)
+{
+	struct shm_service_tbl_ent *entry;
+
+	if (unlikely(*nxt_reqtok >= SHM_BM_SERVICE_TABLE_SIZE)) return -1;
+	entry = &srv_tbl[*nxt_reqtok];
+
+	if (unlikely(memmgr_shared_page_map_aligned(shm_id, SHM_BM_ALIGN, (vaddr_t *) &entry->shmptr)) == 0) return -1;
+	entry->authtok = cos_inv_token();
+	return (*nxt_reqtok)++;
+}
+
+static inline struct shm_service_tbl_ent *
+__shm_bm_srv_auth(struct shm_service_tbl_ent *srv_tbl, shm_reqtok_t reqtok)
+{
+	struct shm_service_tbl_ent *entry;
+
+	if (unlikely(reqtok >= SHM_BM_SERVICE_TABLE_SIZE || reqtok < 0)) return 0;
+
+	entry = &srv_tbl[reqtok];
+	if (unlikely(entry->authtok != cos_inv_token())) return 0;
+
+	return entry;
+}
+
+static inline void *   
+__shm_bm_srv_take(struct shm_service_tbl_ent *srv_tbl, shm_reqtok_t reqtok, shm_objid_t objid, size_t objsz, unsigned int nobj)
+{
+	struct shm_service_tbl_ent *entry;
+	void                       *shmptr;
+
+	if (unlikely(objid >= nobj)) return 0;
+	if (unlikely((entry = __shm_bm_srv_auth(srv_tbl, reqtok)) == 0)) return 0;
+
+	shmptr = entry->shmptr;
+	/* obj has not been allocated */
+	if (unlikely(*(SHM_BM_REFC(shmptr, nobj) + objid) == 0)) return 0;
+
+	cos_faab(SHM_BM_REFC(shmptr, nobj) + objid, 1);
+
+	return SHM_BM_DATA(shmptr, nobj) + (objid * objsz);
+}
+
+static inline void *   
+__shm_bm_srv_take_norefcnt(struct shm_service_tbl_ent *srv_tbl, shm_reqtok_t reqtok, shm_objid_t objid, size_t objsz, unsigned int nobj)
+{
+	struct shm_service_tbl_ent *entry;
+	void                       *shmptr;
+
+	if (unlikely(objid >= nobj)) return 0;
+	if (unlikely((entry = __shm_bm_srv_auth(srv_tbl, reqtok)) == 0)) return 0;
+
+	shmptr = entry->shmptr;
+	// obj has not been allocated
+	if (unlikely(*(SHM_BM_REFC(shmptr, nobj) + objid) == 0)) return 0;
```

Comment:
I don't think that this check actually has much utility. It is a subtle argument so we should chat about it at some point.

Reply:
I thought this would be required since it is possible for a component to take a reference to an unallocated object otherwise. This might not be something we are worried about tho.